File: LeetCode/0437-path-sum-iii/0437-path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int solve(TreeNode root, long curr, int target, Map<Long, Integer> mp) {
        if (root == null) {
            return 0;
        }
        curr += root.val;
        int count = 0;
        if (mp.containsKey(curr - target)) {
            count = mp.get(curr - target);
        }
        mp.put(curr, mp.getOrDefault(curr, 0) + 1);
        count += solve(root.left, curr, target, mp);
        count += solve(root.right,curr, target, mp);
        mp.put(curr, mp.getOrDefault(curr, 0) - 1);
        return count;
    }
    public int pathSum(TreeNode root, int targetSum) {
        Map<Long, Integer> mp = new HashMap<>();   
        mp.put(0L, 1);
        return solve(root, 0L, targetSum, mp);     
    }
}

File: LeetCode/0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root, curr, target, mp) -> int:
        if not root:
            return 0
        curr += root.val
        count = 0
        if curr - target in mp:
            count = mp[curr - target]
        if curr in mp:
            mp[curr] += 1
        else:
            mp[curr] = 1
        count += self.solve(root.left, curr, target, mp)
        count += self.solve(root.right,curr, target, mp)
        mp[curr] -= 1
        return count

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        mp = {}
        mp[0] = 1
        return self.solve(root, 0, targetSum, mp)